File: src/adventofcode2018/Day07.java
package adventofcode2018;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.TreeSet;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import utils.Input;

/**
 * https://adventofcode.com/2018/day/7
 */
public class Day07 {

    static Pattern p = Pattern.compile("Step (.) must be finished before step (.) can begin.");

    public static void main(String[] args) {

        Map<String, List<String>> order = new HashMap<>();
        Set<String> steps = new TreeSet<>();

        List<String> input = new Input(2018, "input07.txt").strings();
        for (String s : input) {
            Matcher m = p.matcher(s);
            m.find();
            String s1 = m.group(1);
            String s2 = m.group(2);
            steps.add(s1);
            steps.add(s2);
            List<String> list;
            if (order.containsKey(s2)) {
                list = order.get(s2);
            } else {
                list = new ArrayList<>();
                order.put(s2, list);
            }
            list.add(s1);
        }
        order.values().stream().forEach(list -> list.sort(Comparator.naturalOrder()));
        System.out.println(order);

        /*
        while (!steps.isEmpty()) {
            String first = steps.stream()
                    .filter(s -> !order.containsKey(s) || order.get(s).isEmpty())
                    .findFirst().get();
            steps.remove(first);
            System.err.print(first);
            order.values().stream().forEach(list -> list.remove(first));
        }
        //*/

        int n = 5;
        int time = 0;
        int workers[] = new int[n];
        String letters[] = new String[n];
        while (!order.isEmpty()) {
            time++;
            for (int i=0; i<n; i++) {
                workers[i] = Math.max(0, workers[i]-1);
            }
            for (int i=0; i<n; i++) {
                if (workers[i] == 0 && letters[i] != null) {
                    String l = letters[i];
                    order.values().stream().forEach(list -> list.remove(l));
                    order.remove(l);
                    letters[i] = null;
                }
            }
            for (int i=0; i<n; i++) {
                if (workers[i] == 0) {
                    Optional<String> first = steps.stream()
                            .filter(s -> !order.containsKey(s) || order.get(s).isEmpty())
                            .findFirst();
                    int ii = i;
                    first.ifPresent(l -> {
                        steps.remove(l);
                        workers[ii] = l.codePointAt(0) - 'A' + 61;
                        letters[ii] = l;
                    });
                }
            }
        }
        System.err.println(time - 1);
    }

}

File: src/Python/2018/Day07.py

from utils import Input, printResult
import copy

# https://adventofcode.com/2018/day/7

pattern = "Step (.) must be finished before step (.) can begin."
steps = set()
order = {}
for m in Input(2018, 7).match(pattern):
    s1, s2 = m[1], m[2]
    steps.add(s1)
    steps.add(s2)
    if s2 not in order:
        order[s2] = []
    order[s2].append(s1)
steps = list(sorted(steps))

def simulate(order, steps, n):
    time, sequence = 0, ""
    workers = [0] * n
    letters = [None] * n
    while len(order) != 0:
        time += 1
        for i in range(n):
            workers[i] = max(0, workers[i] - 1)
            if workers[i] == 0:
                if letters[i]:
                    l = letters[i]
                    for list in order.values():
                        if l in list: list.remove(l)
                    order.pop(l, None)
                    letters[i] = None
                x = next(filter(lambda s: s not in order or len(order[s]) == 0, steps), None)
                if x:
                    steps.remove(x)
                    sequence += x
                    workers[i] = ord(x) - ord('A') + 61
                    letters[i] = x
            
    return (time - 1, sequence)

printResult(1, simulate(copy.deepcopy(order), list(steps), 1)[1])
printResult(2, simulate(copy.deepcopy(order), list(steps), 5)[0])
